txpool/txpool.go

```diff
@@ -823,466 +531,508 @@ func (t *TxPool) validateTx(
 
-	reinsertFunc()
+	// lock promoted
+	account.promoted.lock(true)
+	defer account.promoted.unlock()
 
-	return promote
-}
+	// prune promoted
+	pruned := account.promoted.prune(nonce)
 
-// Peek returns the lowest nonce transaction in the account based heap
-func (t *txHeapWrapper) Peek() *types.Transaction {
-	return t.txs.Peek()
-}
+	// update pool state
+	p.index.remove(pruned...)
+	p.gauge.decrease(slotsRequired(pruned...))
 
-// Push adds a transaction to the account based heap
-func (t *txHeapWrapper) Push(tx *types.Transaction) {
-	// Check if the current transaction has a higher or equal nonce
-	// than all the current transactions in the account based heap
-	i := sort.Search(len(t.txs), func(i int) bool {
-		return t.txs[0].Nonce >= tx.Nonce
-	})
-
-	// If sort.Search found something, it will return the index
-	// of the first found element for which func(i int) was true
-	if i < len(t.txs) && t.txs[i].Nonce == tx.Nonce {
-		// i is an index corresponding to an element in the
-		// account based heap, and the nonces match up, so this tx is discarded
+	if nonce <= account.getNonce() {
+		// only the promoted queue needed pruning
 		return
 	}
 
-	// All checks have passed, add the tx to the account based heap
-	heap.Push(&t.txs, tx)
-}
+	// lock enqueued
+	account.enqueued.lock(true)
+	defer account.enqueued.unlock()
 
-// Pop removes the min-nonce transaction from the account based heap
-func (t *txHeapWrapper) Pop() *types.Transaction {
-	res := heap.Pop(&t.txs)
-	if res == nil {
-		return nil
-	}
+	// prune enqueued
+	pruned = account.enqueued.prune(nonce)
 
-	return res.(*types.Transaction)
-}
+	// update pool state
+	p.index.remove(pruned...)
+	p.gauge.decrease(slotsRequired(pruned...))
 
-// Remove removes the transaction with given hash
-func (t *txHeapWrapper) Remove(hash types.Hash) bool {
-	for i, tx := range t.txs {
-		if tx.Hash == hash {
-			t.txs = append(t.txs[:i], t.txs[i+1:]...)
-			return true
-		}
+	// update next nonce
+	account.setNonce(nonce)
+
+	if first := account.enqueued.peek(); first != nil &&
+		first.Nonce == nonce {
+		// first enqueued tx is expected -> signal promotion
+		p.promoteReqCh <- promoteRequest{account: addr}
 	}
-	return false
 }
 
-// Account based heap implementation //
-// The heap is min-nonce ordered //
+// createAccountOnce is used when discovering an address
+// of a received transaction for the first time.
+// This function ensures that the account
+// is created atomically and only once.
+func (p *TxPool) createAccountOnce(newAddr types.Address) *account {
+	// fetch nonce from state
+	stateRoot := p.store.Header().StateRoot
+	stateNonce := p.store.GetNonce(stateRoot, newAddr)
 
-type txHeap []*types.Transaction
+	// initialize the account
+	account := p.accounts.initOnce(newAddr, stateNonce)
 
-// Required method definitions for the standard golang heap package
+	return account
+}
 
-func (t *txHeap) Peek() *types.Transaction {
-	if len(*t) == 0 {
-		return nil
+/* QUERY methods */
```

Comment:
Feel free to move all the query methods to query.go

Reply:
[Exported](https://github.com/0xPolygon/polygon-sdk/pull/287/commits/8f8c2a793b02c4744e33f3598b871ea3b1aa73e9)